njalla/resource_server.go

```diff
@@ -0,0 +1,177 @@
+package njalla
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/Sighery/gonjalla"
+)
+
+func resourceServer() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceServerCreate,
+		ReadContext:   resourceServerRead,
+		UpdateContext: resourceServerUpdate,
+		DeleteContext: resourceServerDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:		schema.TypeString,
+				Required:	true,
+				Description:	"Name for the server",
+			},
+			"instance_type": {
+				Type:		schema.TypeString,
+				Required:	true,
+				Description:	"Instance type for the server",
+			},
+			"os": {
+				Type:		schema.TypeString,
+				Required:	true,
+				Description:	"OS type for the server",
+			},
+			"public_key": {
+				Type:		schema.TypeString,
+				Required:	true,
+				Description:	"Public key material for this server",
+			},
+			"months": {
+				Type:		schema.TypeInt,
+				Required:	true,
+				Description:	"Number of months to buy the server for",
+			},
+			"public_ip": {
+				Type:		schema.TypeString,
+				Computed:	true,
+				Description:	"Public IPv4 address of this server",
+			},
+
+		},
+
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceServerImport,
+		},
+	}
+}
+
+func resourceServerCreate(
+	ctx context.Context, d *schema.ResourceData, m interface{},
+) diag.Diagnostics {
+	config := m.(*Config)
+
+	name := d.Get("name").(string)
+	instanceType := d.Get("instance_type").(string)
+	os := d.Get("os").(string)
+	publicKey := d.Get("public_key").(string)
+	months := d.Get("months").(int)
+
+	server, err := gonjalla.AddServer(config.Token, name, instanceType, os, publicKey, months)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceServerRead(ctx, d, m)
+}
+
+func resourceServerRead(
+	ctx context.Context, d *schema.ResourceData, m interface{},
+) diag.Diagnostics {
+	config := m.(*Config)
+
+	var diags diag.Diagnostics
+
+	servers, err := gonjalla.ListServers(config.Token)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	for true {
+		for _, server := range servers {
+			if d.Id() == server.ID {
+				if len(server.Ips) == 0 {
+					time.Sleep(5 * time.Second)
```

Comment:
I'll need to look up how providers implement timeouts. Not sure this is the way to go. I know the AWS Terraform provider has different timeouts for different resources, where some resources can take many minutes to read (either when planning, or after applying). I worry that just implementing a `time.Sleep` here will hang execution forever if for some reason the Njalla API fails to ever return the expected value.

My guess is either the SDK has some fancy sleep function for them to keep track of time spent per resource, or will kill any operation running for longer than X time by default.